src/hotspot/share/opto/addnode.cpp

```diff
@@ -872,20 +863,38 @@ const Type *OrLNode::add_ring( const Type *t0, const Type *t1 ) const {
   return TypeLong::make( r0->get_con() | r1->get_con() );
 }
 
+//---------------------------Helper -------------------------------------------
+/* Decide if the given node is used only in arithmetic expressions(add or sub).
+ */
+static bool is_used_in_only_arithmetic(Node* n, BasicType bt) {
+  for (DUIterator_Fast imax, i = n->fast_outs(imax); i < imax; i++) {
+    Node* u = n->fast_out(i);
+    if (u->Opcode() != Op_Add(bt) && u->Opcode() != Op_Sub(bt)) {
+      return false;
+    }
+  }
+  return true;
+}
+
 //=============================================================================
 //------------------------------Idealize---------------------------------------
 Node* XorINode::Ideal(PhaseGVN* phase, bool can_reshape) {
   Node* in1 = in(1);
   Node* in2 = in(2);
-  int op1 = in1->Opcode();
-  // Convert ~(x+c) into (-c-1)-x. Note there isn't a bitwise not
-  // bytecode, "~x" would typically represented as "x^(-1)", so ~(x+c)
-  // will eventually be (x+c)^-1.
-  if (op1 == Op_AddI && phase->type(in2) == TypeInt::MINUS_1 &&
-      in1->in(2)->Opcode() == Op_ConI) {
-    jint c = phase->type(in1->in(2))->isa_int()->get_con();
-    Node* neg_c_minus_one = phase->intcon(java_add(-c, -1));
-    return new SubINode(neg_c_minus_one, in1->in(1));
+
+  // Convert ~x into -1-x when ~x is used in an arithmetic expression
+  // or x itself is an expression.
+  if (phase->type(in2) == TypeInt::MINUS_1) { // follows LHS^(-1), i.e., ~LHS
+    if (phase->is_IterGVN()) {
+      if (is_used_in_only_arithmetic(this, T_INT)
+          // used only once and contains arithmetic
+          || (outcnt() == 1 && (in1->Opcode() == Op_AddI || in1->Opcode() == Op_SubI))) {
```

Comment:
May I ask why do you need `outcnt() == 1` here, thanks.

Reply:
I was thinking of which cases we still want to transform even when the `Xornode` is used in a non-arithmetic expression, and the only case I think safe to transform is that when it is used only once and has some arithmetic inside, for example,
`return ~(x+c)`.

Reply:
In that case, you would need to check the `Opcode` of the use similar to the above line. However, I believe that since you have already deferred the transformation, the bitwise transformations should have been finished by then so you don't need to worry about that. I would suggest adding a comment to remind this. And some tests to confirm, such as `(x + y) & ~(x + y) == 0`. Thanks.

Reply:
Thanks. I agree transformation like `(x + y) & ~(x + y) => 0` should have been covered by idealization from `AndINode` at this point. However, my intention was to catch the case like `return ~(x-1) => return -x`. If we have only line 889 we would not be able to do this because `~(x - 1)` is not used in arithmetic and will not be transformed.

Reply:
I mean you don't need `outcnt() == 1`.